src/sinks/datadog/traces/apm_stats/flusher.rs

```diff
@@ -0,0 +1,173 @@
+use std::{
+    io::Write,
+    sync::{Arc, Mutex},
+};
+
+use bytes::Bytes;
+use snafu::ResultExt;
+use tokio::sync::oneshot::{Receiver, Sender};
+use vector_common::{finalization::EventFinalizers, request_metadata::RequestMetadata};
+
+use super::{
+    aggregation::Aggregator, build_request, DDTracesMetadata, DatadogTracesEndpoint,
+    DatadogTracesEndpointConfiguration, RequestBuilderError, StatsPayload,
+    BUCKET_DURATION_NANOSECONDS,
+};
+use crate::{
+    http::{BuildRequestSnafu, HttpClient},
+    sinks::util::{Compression, Compressor},
+};
+
+/// Flushes cached APM stats buckets to Datadog on a 10 second interval.
+/// When the sink signals this thread that it is shutting down, all remaining
+/// buckets are flush before the thread exits.
+///
+/// # arguments
+///
+/// * `tripwire`                 - Receiver that the sink signals when shutting down.
+/// * `client`                   - HttpClient to use in sending the stats payloads.
+/// * `compression`              - Compression to use when creating the HTTP requests.
+/// * `endpoint_configuration`   - Endpoint configuration to use when creating the HTTP requests.
+/// * `aggregator`               - The Aggregator object containing cached stats buckets.
+pub async fn flush_apm_stats_thread(
+    mut tripwire: Receiver<Sender<()>>,
+    client: HttpClient,
+    compression: Compression,
+    endpoint_configuration: DatadogTracesEndpointConfiguration,
+    aggregator: Arc<Mutex<Aggregator>>,
+) {
+    let sender = ApmStatsSender {
+        client,
+        compression,
+        endpoint_configuration,
+        aggregator,
+    };
+
+    // flush on the same interval as the stats buckets
+    let mut interval =
+        tokio::time::interval(std::time::Duration::from_nanos(BUCKET_DURATION_NANOSECONDS));
+
+    debug!("Starting APM stats flushing thread.");
+
+    loop {
+        tokio::select! {
+
+        _ = interval.tick() => {
+            // flush the oldest bucket from the cache to Datadog
+            sender.flush_apm_stats(false).await;
+        },
+        signal = &mut tripwire =>  match signal {
+            // sink has signaled us that the process is shutting down
+            Ok(sink_shutdown_ack_sender) => {
+
+                debug!("APM stats flushing thread received exit condition. Flushing remaining stats before exiting.");
+                sender.flush_apm_stats(true).await;
+
+                // signal the sink (who tripped the tripwire), that we are done flushing
+                let _ = sink_shutdown_ack_sender.send(());
+                break;
+            }
+            Err(_) => error!(message = "Tokio Sender unexpectedly dropped.", internal_log_rate_limit = true),
```

Comment:
I think we should also be breaking here? If the tripwire signal is terminal (i.e. we `break`) when sent correctly, we would conceivably want to shutdown if the tripwire channel was closed: it's not like it's gonna actually ever _come back_.

Reply:
10000% yes. Good catch!